tests/testthat/test-statistical_correctness.R

```diff
@@ -0,0 +1,408 @@
+  r0_values <- c(1.9, 2.0, 4.0, 12.0)
+
+  for (r0 in r0_values) {
+    epi_outcome_iterative <- final_size(
+      r0 = r0,
+      contact_matrix = contact_matrix,
+      demography_vector = demography_vector,
+      susceptibility = susceptibility,
+      p_susceptibility = p_susceptibility,
+      solver = "iterative",
+      control = list(
+        tolerance = 1e-4 # Newton solver requires lower tolerance
+      )
+    )
+
+    epi_outcome_newton <- final_size(
+      r0 = r0,
+      contact_matrix = contact_matrix,
+      demography_vector = demography_vector,
+      susceptibility = susceptibility,
+      p_susceptibility = p_susceptibility,
+      solver = "newton",
+      control = list(
+        tolerance = 1e-5 # Newton solver requires lower tolerance
+      )
+    )
+
+    # check for correct answer
+    tolerance <- 1e-4
+    expect_equal(
+      epi_outcome_iterative$p_infected, epi_outcome_newton$p_infected,
+      tolerance = tolerance,
+      info = sprintf("Failed for R0 = %f", r0)
+    )
+  }
+})
+
+#### Check solver equivalence with POLYMOD data ####
+
+# Prepare common elements for testing
+polymod <- socialmixr::polymod
+contact_data <- socialmixr::contact_matrix(
+  polymod,
+  countries = "United Kingdom",
+  age.limits = c(0, 20, 40),
+  symmetric = TRUE
+)
+contact_matrix <- contact_data$matrix
+demography_vector <- contact_data$demography$population
+
+# scale by maximum real eigenvalue and divide by demography
+contact_matrix <- contact_matrix / max(eigen(contact_matrix)$values)
+contact_matrix <- contact_matrix / demography_vector
+
+# Check final_size works with Newton solver
+test_that("Check finalsize by groups works for Polymod, newton solver", {
+  r0 <- 2.0
+
+  n_demo_grps <- length(demography_vector)
+  n_risk_grps <- 4
+
+  # prepare p_susceptibility and susceptibility
+  psusc <- matrix(
+    data = 1, nrow = n_demo_grps, ncol = n_risk_grps
+  )
+  psusc <- psusc / rowSums(psusc)
+  susc <- matrix(
+    data = 1, nrow = n_demo_grps, ncol = n_risk_grps
+  )
+
+  # prepare control
+  control <- list(
+    iterations = 10000,
+    tolerance = 1e-6
+  )
+
+  # prepare outcome with iterative solver
+  epi_outcome_iterative <- final_size(
+    r0 = r0,
+    contact_matrix = contact_matrix,
+    demography_vector = demography_vector,
+    p_susceptibility = psusc,
+    susceptibility = susc,
+    solver = "iterative",
+    control = control
+  )
+
+  # prepare outcome with newton solver
+  epi_outcome_newton <- final_size(
+    r0 = r0,
+    contact_matrix = contact_matrix,
+    demography_vector = demography_vector,
+    p_susceptibility = psusc,
+    susceptibility = susc,
+    solver = "newton",
+    control = control
+  )
+
+  # check for equivalence
+  expect_equal(
+    epi_outcome_iterative$p_infected,
+    epi_outcome_newton$p_infected,
+    tolerance = 1e-5
+  )
+})
+
+#### Check for correct final size calculation in complex data case ####
+
+# test taken from EvL
```

Comment:
It would be useful at some point if you manage to find out where these values come from. It would be a good proof if it comes from a paper which computes the finalsize from a different implementation.

Reply:
The only information I've been able to find is this set of comments in @BlackEdder's code:

> ```
> // These were generated using:
> // cm_lst <- contactsr::load_polymod_data(age_limits=c(15,30,45,60,75))
> // contactsr::symmetric_matrix(cm_lst)
> ```

I have no idea to what extent this method is used or reported in the literature - or where to look - would be very useful to be pointed towards these sources.

Reply:
> Thanks, I think it's already much clearer!
> 
> There is still some overlap, which I flagged for the most part. I wonder if splitting tests across files is not making things more difficult and cause this redundancy.
> 
> It would also be useful to have snapshot tests somewhere to detect possible regressions in the future.

Thanks - I've removed an old testing file, and removed some extra tests as well. I've removed the Newton solver test that was also for statistical correctness, and instead included a similar test under `test-statistical_correctness.R`.

Reply:
Snapshot tests ensure future changes don't modify the output. It's pretty crucial, even if the values don't come from an external source, as the existing tests only check the range (as you mention) and thus leave quite a lot of room for the errors.

If tomorrow, I make a change which modifies each value by 5%, it's very likely it won't be detected by the current suite of tests and it's something we want to avoid.

To clarify, I agree with you that this kind of snapshot tests would not add anything for statistical correctness checking but it doesn't mean they're useless altogether.

Reply:
I agree with the principle, but for this test where we check the range, maybe the values should be allowed to vary by about 5% - the range [0.3, 0.45] is actually quite large. I can add a snapshot for better certainty as well.

Reply:
Yes, I agree. It's one of the reason why it would be useful to understand where this range comes from

Reply:
Pinging @BlackEdder who hopefully knows?

Reply:
For now, I've added snapshots for both solver in https://github.com/epiverse-trace/finalsize/pull/106/commits/cdcb701da4d76273369a4357a6200c4d175b974c. I do notice that they are not the same values, although the test for solver equivalence https://github.com/epiverse-trace/finalsize/blob/e76868b0950bf84d51e0242b1bd8302cc81425e7/tests/testthat/test-statistical_correctness.R#L258 passes fine. If this is an adequate solution for now, I will mark this as resolved.

Reply:
Aren't the differences just due to the fact you changed the demography vector for the iterative solver on L373?

This looks fine to me :+1:

Reply:
Yes, thanks - that was the reason.